src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets

```diff
@@ -2010,56 +1990,37 @@ because xbuild doesn't support framework reference assemblies.
 </Target>
 
-<Target Name="_CopyPdbFiles"
-		Inputs="@(_ResolvedPortablePdbFiles)"
-		Outputs="$(_AndroidStampDirectory)_CopyPdbFiles.stamp"
-		DependsOnTargets="_ConvertPdbFiles">
-	<ItemGroup>
-		<_CopyPdbFilesLinkerFiles Include="@(_ResolvedPortablePdbFiles->'$(MonoAndroidLinkerInputDir)%(Filename)%(Extension)')" />
-	</ItemGroup>
-	<CopyIfChanged SourceFiles="@(_ResolvedPortablePdbFiles)" DestinationFiles="@(_CopyPdbFilesLinkerFiles)" />
-	<Touch Files="$(_AndroidStampDirectory)_CopyPdbFiles.stamp" AlwaysCreate="True" />
-	<ItemGroup>
-		<FileWrites Include="@(_CopyPdbFilesLinkerFiles)" />
-	</ItemGroup>
-</Target>
-
-<Target Name="_CopyMdbFiles"
-		Inputs="@(_ResolvedMdbFiles)"
-		Outputs="$(_AndroidStampDirectory)_CopyMdbFiles.stamp"
-		DependsOnTargets="_CollectMdbFiles">
-	<ItemGroup>
-		<_CopyMdbFilesLinkerFiles Include="@(_ResolvedMdbFiles->'$(MonoAndroidLinkerInputDir)%(Filename)%(Extension)')" />
-	</ItemGroup>
-	<CopyIfChanged SourceFiles="@(_ResolvedMdbFiles)" DestinationFiles="@(_CopyMdbFilesLinkerFiles)" />
-	<Touch Files="$(_AndroidStampDirectory)_CopyMdbFiles.stamp" AlwaysCreate="True" />
-	<ItemGroup>
-		<FileWrites Include="@(_CopyMdbFilesLinkerFiles)" />
-	</ItemGroup>
-</Target>
-	
 <Target Name="_LinkAssemblies"
   DependsOnTargets="_ResolveAssemblies;_CreatePackageWorkspace;$(_BeforeLinkAssemblies);_GenerateJniMarshalMethods;_LinkAssembliesNoShrink;_LinkAssembliesShrink" />
 
 <Target Name="_GenerateJniMarshalMethods"
     Condition="'$(AndroidGenerateJniMarshalMethods)' == 'True' And '$(AndroidLinkMode)' != 'None' And '$(OS)' != 'Windows_NT'"
     DependsOnTargets="_GetReferenceAssemblyPaths;_SetLatestTargetFrameworkVersion"
-    Inputs="@(ResolvedUserAssemblies->'$(MonoAndroidLinkerInputDir)%(Filename)%(Extension)')"
-    Outputs="$(_AndroidJniMarshalMethodsFlag)">
+    Inputs="$(_AndroidBuildPropertiesCache);@(ResolvedUserAssemblies)"
+    Outputs="$(_AndroidStampDirectory)_GenerateJniMarshalMethods.stamp">
+  <PropertyGroup>
+    <_JniMarshalMethodsInputDir>$(IntermediateOutputPath)jnisrc\</_JniMarshalMethodsInputDir>
+  </PropertyGroup>
   <ItemGroup>
     <_JniFrameworkAssembly Include="Mono.Android.dll" />
     <_JniFrameworkAssembly Include="OpenTK-1.0.dll" />
     <_JniFrameworkAssembly Include="OpenTK.dll" />
     <_JniFrameworkAssembly Include="Xamarin.Android.NUnitLite.dll" />
-  </ItemGroup>
-  <ItemGroup>
     <_AssembliesToProcess Include="@(ResolvedUserAssemblies)" />
-    <_AssembliesToProcess Include="@(ResolvedAssemblies)" Condition=" '%(Filename)' != '' And '@(_JniFrameworkAssembly)' != '' " />
+    <_AssembliesToProcess Include="@(ResolvedFrameworkAssemblies)" Condition=" '%(Filename)' == '@(_JniFrameworkAssembly->'%(Filename)')' " />
+    <_JniMarshalMethodAssemblies Include="@(ResolvedAssemblies->'$(_JniMarshalMethodsInputDir)%(Filename)%(Extension)')" />
   </ItemGroup>
+  <CopyIfChanged
+      SourceFiles="@(ResolvedAssemblies)"
+      DestinationFiles="@(_JniMarshalMethodAssemblies)"
+  />
```

Comment:
We can avoid copying the assemblies by using `-o <output directory>` option when running the `jnimarshalmethod-gen.exe`. That should speed it up a bit more as well.

Reply:
The only problem would be that `MONO_PATH` would then need to include all the directories from `@(ResolvedAssemblies)`, these could be lots of paths such as `~/.nuget/packages/xamarin.forms/...`, etc.

I think I could convert this `<Exec/>` call to a proper `<GenerateJniMarshalMethods/>` MSBuild task to do it in a nice way.

Do you think we could do that down the road?

Reply:
@jonathanpeppers the problem with that is the `jnimarshalmethod-gen.exe` to load the monodroid assemblies (like System.dll etc) and if we turn this into a full task, then it will use the .net assemblies already loaded by the build system. 

Unless you plan to make this a ToolTask? In which case that might work.

Reply:
I think the `MONO_PATH` might just need to contain our framework assemblies path (plus facades) and not the user referenced assemblies folders. What error do you get when using the `-o` option?

Reply:
@radekdoulik the error is:

```
Parent class failed to load, due to: Generic Type Definition failed to init, due to: Could not load file or assembly 'Mono.Android, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065' or one of its dependencies. assembly:/Users/jonathanpeppers/Desktop/Git/xamarin-android/tests/TestRunner.Core/bin/Release/TestRunner.Core.dll type:TestInstrumentation`1 member:(null) assembly:/Users/jonathanpeppers/Desktop/Git/xamarin-android/tests/TestRunner.Core/bin/Release/TestRunner.Core.dll type:TestInstrumentation`1 member:(null)
  System.TypeLoadException: Parent class failed to load, due to: Generic Type Definition failed to init, due to: Could not load file or assembly 'Mono.Android, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065' or one of its dependencies. assembly:/Users/jonathanpeppers/Desktop/Git/xamarin-android/tests/TestRunner.Core/bin/Release/TestRunner.Core.dll type:TestInstrumentation`1 member:(null) assembly:/Users/jonathanpeppers/Desktop/Git/xamarin-android/tests/TestRunner.Core/bin/Release/TestRunner.Core.dll type:TestInstrumentation`1 member:(null)
    at (wrapper managed-to-native) System.RuntimeType.GetMethodsByName_native(System.RuntimeType,intptr,System.Reflection.BindingFlags,System.RuntimeType/MemberListType)
    at System.RuntimeType.GetMethodsByName (System.String name, System.Reflection.BindingFlags bindingAttr, System.RuntimeType+MemberListType listType, System.RuntimeType reflectedType) [0x0001b] in <e026bace0e4b47ea9d52101abff29ed0>:0
    at System.RuntimeType.GetMethodCandidates (System.String name, System.Reflection.BindingFlags bindingAttr, System.Reflection.CallingConventions callConv, System.Type[] types, System.Int32 genericParamCount, System.Boolean allowPrefixLookup) [0x00010] in <e026bace0e4b47ea9d52101abff29ed0>:0
    at System.RuntimeType.GetMethods (System.Reflection.BindingFlags bindingAttr) [0x00000] in <e026bace0e4b47ea9d52101abff29ed0>:0
    at Xamarin.Android.Tools.JniMarshalMethodGenerator.App.CreateMarshalMethodAssembly (System.String path) [0x00286] in /Users/jonathanpeppers/Desktop/Git/xamarin-android/external/Java.Interop/tools/jnimarshalmethod-gen/App.cs:347
    at Xamarin.Android.Tools.JniMarshalMethodGenerator.App.ProcessAssemblies (System.Collections.Generic.List`1[T] assemblies) [0x00150] in /Users/jonathanpeppers/Desktop/Git/xamarin-android/external/Java.Interop/tools/jnimarshalmethod-gen/App.cs:186
```

`$(_XATargetFrameworkDirectories)` is `xamarin-android/bin/Debug/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v1.0/`.

I think this happens because Mono.Cecil is trying to load `Mono.Android.dll` that is referenced by `TestRunner.Core.dll`. I think it worked before because they all got copied to a single directory: either `linksrc` or this new `jnisrc` directory.

So I reworked things a bit more to fill out `MONO_PATH`, then `<LinkAssemblies/>` fails because some of the assemblies don't make it to the `jnisrc` directory.

Here is how far I got: https://github.com/xamarin/xamarin-android/commit/96cc777d28f662a644b5270c275ebbb8a8b9f2a7

I'll look into it some more tomorrow.

Reply:
> Unless you plan to make this a ToolTask? In which case that might work.

@dellis1972 yes it would be a `ToolTask`.

I started down that path when I first removed `linksrc`:

https://github.com/jonathanpeppers/xamarin-android/blob/cc37f9d38abc0160ec869b88cbddf1c0a047eccd/src/Xamarin.Android.Build.Tasks/Tasks/JniMarshalMethodGen.cs#L12

But I don't know if we need it yet.

Reply:
The above error doesn't come from cecil, it is from reflection. `jnimarshalmethod-gen` has to do both.

We might try to add a new option to `jnimarshalmethod-gen` to let it know about all the references and preload them, before doing the reflection work. The same information might be used for cecil part, to prefill the resolver.

Reply:
About the ToolTask, we already had that as part of https://github.com/xamarin/xamarin-android/pull/2454/files, which didn't land because of the mono on Windows issues and is probably quite outdated now :-(

Reply:
That does bring up another question, should we make `jnimarshalmethod-gen` use a response file? I have a feeling we will hit command line length or path length issues with this stuff, so we should look at keeping those as short a possible (and do the same for any future command line tooling we produce).

Reply:
That would make sense. Fortunately we use `Mono.Options`, so it should hopefully be relatively easy.

Reply:
Prepared https://github.com/xamarin/java.interop/pull/465 to support response files and additional `-r` option.

I have also tried https://github.com/xamarin/xamarin-android/commit/96cc777d28f662a644b5270c275ebbb8a8b9f2a7 and `jnimarshalmethod-gen` worked OK with locale and xforms tests, till the build failed in the linker task (some of the references were not resolved?).

@jonathanpeppers Which test or app did you use to get the error mentioned yesterday? (I would like to test it with the new response file and ref options)

Reply:
I was just building `Mono.Android-Tests.csproj` in `Release`.

The linker is still going to fail because assemblies are missing. So it's going to have to do some `Exists()` check and use `@(ResolvedAssemblies)` in that case...